app/master/build_scheduler.py

```diff
@@ -49,7 +49,13 @@ def needs_more_slaves(self):
 
         :rtype: bool
         """
-        return self._num_executors_allocated < self._max_executors and not self._build._unstarted_subjobs.empty()
+        if self._num_executors_allocated >= self._max_executors:
+            return False
+        if self._build._unstarted_subjobs.empty():
+            return False
+        if self._num_executors_allocated >= len(self._build.all_subjobs()):
```

Comment:
In the scenario where the slaves being allocated are trickling in (being released from other builds onto this one), then the length of all subjobs is an overestimate, whereas the unstarted_subjobs.qsize() is more of an exact fit.

It's really not a big deal though. I'm happy with leaving it as is and merging it.